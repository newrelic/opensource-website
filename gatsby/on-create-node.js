// https://www.gatsbyjs.org/docs/actions/#createNode

const { v4: uuidv4 } = require('uuid');

const isMdx = (type) => type === 'Mdx';
const isProject = (type) => type === 'Projects';

const slugFromAbsoluteFilePath = (rootDir, fileAbsolutePath) => {
  // Directory
  if (fileAbsolutePath.indexOf('index.mdx') >= 0) {
    const start = fileAbsolutePath.indexOf(rootDir) + rootDir.length;
    const end = fileAbsolutePath.length - '/index.mdx'.length;
    const slug = fileAbsolutePath.slice(start, end);
    return slug;
  }

  // Single file
  const slug = fileAbsolutePath
    .slice(
      fileAbsolutePath.indexOf(rootDir) + rootDir.length,
      fileAbsolutePath.length
    )
    .replace('.mdx', '');
  return slug;
};

const createProjectMainContent = ({ node, actions }) => {
  const { createNode } = actions;
  const fileAbsolutePath =
    node.internal.type === 'MarkdownRemark'
      ? node.fileAbsolutePath
      : node.internal.contentFilePath;
  const projectMainContentDir = 'src/data/project-main-content/';
  const isMainContent = fileAbsolutePath.indexOf(projectMainContentDir) >= 0;

  if (!isMainContent) {
    return;
  }
  const slug = slugFromAbsoluteFilePath(
    projectMainContentDir,
    fileAbsolutePath
  );

  const fieldData = {
    slug,
  };

  createNode({
    ...fieldData,

    // Required fields.
    id: uuidv4(),
    parent: node.id,
    children: [],
    internal: {
      type: `ProjectMainContent`,
      contentDigest: node.internal.contentDigest, // We can safely reuse this hash since it's content-based
    },
  });
};

module.exports = (params) => {
  const { node } = params;

  if (isProject(node.internal.type)) {
    if (!node.projectType) {
      node.projectType = 'newrelic';
    }
  }

  // Transform readmes into an object type of ProjectMainContent which we'll stitch into GraphQL in `type-defs.gql`
  // This will enable us to get all data needed for a page with one dynamic query
  if (isMdx(node.internal.type)) {
    createProjectMainContent(params);
  }
};
