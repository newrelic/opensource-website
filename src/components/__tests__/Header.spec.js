import React from 'react';
import { render } from '@testing-library/react';
import { useStaticQuery } from 'gatsby';
import {
  LocationProvider,
  createHistory,
  createMemorySource,
} from '@reach/router';
import LocaleProvider from '@newrelic/gatsby-theme-newrelic/src/components/LocaleProvider.mjs';
import themeTranslations from '@newrelic/gatsby-theme-newrelic/src/i18n/translations/en.json';
import i18n from 'i18next';
import { I18nextProvider } from 'react-i18next';
import { QueryClient, QueryClientProvider } from 'react-query';

import Header from '../Header';

const source = createMemorySource('/');
const history = createHistory(source);

const THEME_NAMESPACE = 'gatsby-theme-newrelic';

const queryClient = new QueryClient();

const initI18n = () => {
  i18n.init({
    lng: 'en',
    resources: {
      en: {
        [THEME_NAMESPACE]: themeTranslations,
      },
    },
    defaultNS: 'translation',
    initImmediate: false,
    fallbackLng: 'en',
    ns: [THEME_NAMESPACE, 'translation'],
    interpolation: {
      escapeValue: false,
    },
    react: {
      useSuspense: false,
    },
  });
};

beforeEach(() => {
  useStaticQuery.mockImplementation(() => ({
    allMdx: {
      nodes: [],
    },
    allLocale: {
      nodes: [{ name: 'English', isDefault: true }],
    },
    newRelicThemeConfig: {
      forceTrailingSlashes: false,
    },
    site: {
      siteMetadata: {
        title: 'New Relic Open Source',
        description:
          "New Relic <3's open source. We built this site to make it easy for you to explore the projects we're maintaining as well as our involvement in open standards.",
        author: '@newrelic',
      },
      layout: {
        maxWidth: '1180px',
      },
    },
  }));
});

describe('Header', () => {
  it('renders correctly', () => {
    initI18n();

    const { container } = render(
      <I18nextProvider i18n={i18n}>
        <LocaleProvider i18n={i18n}>
          <QueryClientProvider client={queryClient}>
            <LocationProvider history={history}>
              <Header />
            </LocationProvider>
          </QueryClientProvider>
        </LocaleProvider>
      </I18nextProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
