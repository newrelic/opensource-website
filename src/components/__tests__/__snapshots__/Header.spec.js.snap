// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
Array [
  <div
    className="globalHeaderContainer  "
  >
    <div
      className="globalHeaderContent"
    >
      <div
        className="leftSideLinksContainer"
      >
        <a
          className="logo"
          href="https://newrelic.com/"
          rel="noopener noreferrer"
        >
          New Relic
        </a>
        <ul
          className="leftSideLinks"
        >
          <li
            className="leftSideLinkItem"
          >
            <a
              className="leftSideLink"
              href="https://developer.newrelic.com/"
            >
              Developers
            </a>
          </li>
          <li
            className="leftSideLinkItem active"
          >
            <a
              className="leftSideLink"
              href="#"
            >
              Open Source
            </a>
          </li>
          <li
            className="leftSideLinkItem"
          >
            <a
              className="leftSideLink"
              href="https://docs.newrelic.com/"
            >
              Docs
            </a>
          </li>
        </ul>
      </div>
      <ul
        className="rightSideButtons"
      >
        <li
          className="darkModeToggle"
        />
        <li
          className="rightSideButton githubButton"
        >
          <a
            className="githubButtonLink"
            href="https://github.com/newrelic/opensource-website"
          >
            GitHub
          </a>
        </li>
      </ul>
    </div>
  </div>,
  <header
    className="primaryHeaderContainer  "
  >
    <a
      className="primaryHeaderLogo"
      href="/"
    >
      New Relic Open Source
    </a>
    <ul
      className="primaryHeaderNavLinks"
    >
      <li
        className="primaryHeaderNavLinkItem"
      >
        <a
          className="primaryHeaderNavLink"
          href="/external-projects"
        >
          External Projects
        </a>
      </li>
      <li
        className="primaryHeaderNavLinkItem"
      >
        <a
          className="primaryHeaderNavLink"
          href="/explore-projects"
        >
          New Relic Projects
        </a>
      </li>
    </ul>
    <div
      className="primaryHeaderMobileNav "
    >
      <button
        className="mobileMenuButton"
        onClick={[Function]}
      >
        Menu
      </button>
      <ul
        className="primaryHeaderMobileNavLinks"
      >
        <li
          className="primaryHeaderNavLinkItem"
        >
          <a
            className="primaryHeaderNavLink"
            href="/external-projects"
          >
            External Projects
          </a>
        </li>
        <li
          className="primaryHeaderNavLinkItem"
        >
          <a
            className="primaryHeaderNavLink"
            href="/explore-projects"
          >
            New Relic Projects
          </a>
        </li>
      </ul>
    </div>
  </header>,
]
`;
